### ISSUE TITLE
AnnotatedString equality contradicts docs

### ISSUE THREAD

**Issue by aplavin** [JuliaLang/julia#55244]
Docs:
```julia
  ==(x, y)

  Generic equality operator. Falls back to ===. Should be implemented for all types with a notion of equality, based on the abstract value
  that an instance represents. For example, all numeric types are compared by numeric value, ignoring type. Strings are compared as
  sequences of characters, ignoring encoding. 
<...>

  ==(a::AbstractString, b::AbstractString) -> Bool

  Test whether two strings are equal character by character (technically, Unicode code point by code point).
```

Actual behavior:
```julia
julia> s1 = Base.AnnotatedString("abc def", [(1:1, :x => :y)])
"abc def"

julia> s2 = "abc def"
"abc def"

julia> s1 == s2
false
```

Either the docs or the implementation are clearly wrong...

**Comment by tecosaur** [comment]
In your example `s1` and `s2` are not equal because the first character of `s1` has an annotation. This lines up with "Test whether two strings are equal character by character", but not the "(technically... )" bit.

Initially, I didn't have annotations matter when comparing an `AnnotatedString` with an unannotated string, but that makes string equality non-transitive, which is a big no-no.

**Comment by tecosaur** [comment]
> Not sure how annotations affect code points

They don't, but the reasoning I put forwards is predicated on a character with annotations not being equal to a character without annotations.  

> Sure! I expected == to ignore annotations, following all these explanations in the docs.

We could do this, but I'd expect this behavior to be potentially surprising.

I think the situation can be summed up like so:

## Current behavior

A string without annotations is equal to an unannotated string.

Annotated strings are equal when the underlying string is equal, as are annotations.

### Option 1

Loosen the "technically" to something like "often" in the `==(::AbstractString, ::AbstractString)` docstring.

### Option 2

Ignore annotations entirely when comparing `AnnotatedString`s and other strings with `==`.

**Comment by fatteneder** [comment]
# Option 3

Just document
```
==(a::AnnotatedString, b::AnnotatedString) -> Bool

  Test whether their strings and annotations are equal.
```

**Comment by tecosaur** [comment]
Ah yep, though maybe that would need to explicitly mention

```julia
==(a::AnnotatedString, b::AnnotatedString)
==(a::AbstractString, b::AnnotatedString)
==(a::AnnotatedString, b::AbstractString)
```

**Comment by tecosaur** [comment]
Ahh, I think that the `AnnotatedChar` equality should be updated to match.

**Comment by nhz2** [comment]
### Option 4
Don't make `AnnotatedString` a subtype of `AbstractString`

This way `AnnotatedString` is free to define whatever equality methods it wants without being constrained by the existing `AbstractString` interface.

**Comment by tecosaur** [comment]
Half the point of `AnnotatedString` is that it can be passed to functions that are written for an `AbstractString` and don't know what to do with annotations and it "just works", automatically falling back to a plain un-annotated string :upside_down_face:

**Comment by nhz2** [comment]
The issue is that as a Julia user, I have no idea which functions that use `AbstractString` are going to suddenly have different behavior when passed an `AnnotatedString` because of some strange overload or piracy. If you want a function to have fundamentally different behavior for `AnnotatedString` why not just give the function a new name, for example, instead of `==` define `isequal_annotated(a::AbstractString, b::AbstractString) = a == b && annotations(a) == annotations(b)`

And define a fall back: `annotations(::AbstractString) = Tuple{UnitRange{Int64}, Pair{Symbol, Any}}[]`

**Comment by tecosaur** [comment]
Mmmm, I think there's an argument to be made for abandoning the inclusion of annotations in the concept of string equality, and require them to be checked separately if you do care.

I'm not sure if this is the path we should take, but that is the alternative to the current situation.

**Comment by tecosaur** [comment]
Indeed, this still needs to be worked out, but we've added the "Experimental!" note to give ourselves the space to do so.

**Comment by aplavin** [comment]
Any updates on this?.. AnnotatedString behavior still contradicts the `==` contract:
```julia
help?> ==
<...>
  ==(a::AbstractString, b::AbstractString) -> Bool

  Test whether two strings are equal character by character (technically, Unicode code point by code point).



julia> s1 = StyledStrings.AnnotatedString("abc def", [(1:1, :x => :y)])
julia> s2 = "abc def"

julia> s1 == s2
false
julia> codeunits(s1) == codeunits(s2)
true
```

**Comment by tecosaur** [comment]
I'm still not sure what the right option to go for is, between options 1, 2, and 3 above (I don't think 4 is a good option).

**Comment by adienes** [comment]
IMO option 1 is breaking, and option 3 is incorrect. so process of elimination leaves option 2: ignore annotations.

**Comment by adienes** [comment]
so should we move forward with ignoring annotations?

my (biased) summary of pros and cons being

pros:
* retain transitivity of `==`
* comply with docstring of `==`

cons:
* superficially surprising given that "equal" things appear different

I think the correctness points in the pros outweigh the ergonomics point in cons.

the only way to get all three would require
* removing the ability to `==` across annotated and non-annotated strings, an enormous ergonomics regression
* changing the docstring of specific `==` methods to conflict with the abstract docstring, which is pretty questionable

**Comment by vtjnash** [comment]
No, ignoring annotation is probably most incorrect, since AnnotatedChar is a superset of Char. I think the main issue is with the documentation appearing to claim that AbstractChar represents all unicode code points past present and future even those which haven't been invented yet. However, that is the job of a concretion (e.g. Char) not an abstraction. So that claim is supposed to be a weaker form of normative clarification rather than a prescriptive restriction.

**Comment by adienes** [comment]
I don't see how it can be possible to observe annotations, allow `==(::AnnotatedChar, ::Char)`, and respect transitivity. 

unless the suggestion is to make anything with annotations `!=` to anything without? which would work fine, but sounds like that was previously rejected for usability concerns.

**Comment by vtjnash** [comment]
I think tecosaur mentioned this above: A Char equals an AnnotatedChar if they hold the same codepoint and the same annotations (e.g. no annotations), the same as a sequence of chars (aka a string): https://github.com/JuliaLang/julia/issues/55244#issuecomment-2282666368

**Comment by adienes** [comment]
I agree that would fix equality, but it breaks `convert` which currently strips annotations. then `convert(String, styled_string)` would convert to something not `==` to the input.

currently the only other type that does this that I can think of are floats, which are exceedingly special.

**Comment by aplavin** [comment]
> A Char equals an AnnotatedChar if they hold the same codepoint and the same annotations 

Also, this is unambiguously a breaking change to Julia Base (specifically AbstractChar / AbstractString) contract.

**Comment by vtjnash** [comment]
IIRC, convert does not do that already (eg Int to Float), so there is no issue there

**Comment by JeffBezanson** [comment]
From triage: `==` should compare annotations, and it is not the end of the world for `convert` not to preserve equality, for example `convert(Float64, 1//3) != 1//3`. It's certainly desirable for convert to either preserve equality or error, but convert methods of this other kind don't seriously break anything as long as it does something reasonable.

We need to keep `AnnotatedString <: AbstractString`.

The doc string is correct in one direction: if `==` is true, then the strings iterate the same characters. Doc strings also aren't formal specifications. Also I see the doc string has already been updated :)

**Comment by oscardssmith** [comment]
Triage thinks that as of 1.12.6, the docs and implementation are correct. (although it's somewhat concerning that that PR never linked to this issue), but https://github.com/JuliaLang/julia/pull/55457 should be merged to correct the docs for the Char case.

**Comment by aplavin** [comment]
> Doc strings also aren't formal specifications.

Hmm, how to understand what behavior one can rely on then, if not by docs?
